// zcatalyst-cli/lib/commands/ds/import.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ansi_colors_1 = require("ansi-colors");
const path_1 = require("path");
const endpoints_1 = require("../../endpoints");
const error_1 = __importDefault(require("../../error"));
const command_1 = __importDefault(require("../../internal/command"));
const prompt_1 = __importDefault(require("../../prompt"));
const runtime_store_1 = __importDefault(require("../../runtime-store"));
const constants_1 = require("../../util_modules/constants");
const env_1 = require("../../util_modules/env");
const fs_1 = require("../../util_modules/fs");
const logger_1 = require("../../util_modules/logger");
const option_1 = require("../../util_modules/option");
const shell_1 = require("../../util_modules/shell");
const getWriteConfigFromPath = (pth) => __awaiter(void 0, void 0, void 0, function* () {
    const configJSON = yield fs_1.ASYNC.readJSONFile((0, path_1.resolve)(pth), {
        checkpath: true,
        throws: false
    });
    if (!configJSON || Object.keys(configJSON).length === 0) {
        (0, logger_1.debug)('No parsable import config given');
        return {};
    }
    return configJSON;
});
exports.default = new command_1.default('ds:import [file]')
    .description('Bulk write records to a table in the Catalyst Data Store')
    .option('--table <name|id>', 'Name or ID of the table to which data to be written (e.g. "UserDetails")')
    .option('--config <path>', 'Path of the configuration json file to be used for import.')
    .option('--production', 'flag for pointing to production environment')
    .needs('auth', [constants_1.SCOPE.projects, constants_1.SCOPE.datastore, constants_1.SCOPE.row])
    .needs('config', { optional: true })
    .needs('rc', { optional: true })
    .action((uploadFile) => __awaiter(void 0, void 0, void 0, function* () {
    const env = (0, option_1.getOptionValue)('production', false) ? 'Production' : 'Development';
    const configPath = (0, option_1.getOptionValue)('config', false);
    const importConfig = configPath ? yield getWriteConfigFromPath(configPath) : {};
    const tableId = (0, option_1.getOptionValue)('table', importConfig.table_identifier + '');
    if (!tableId) {
        throw new error_1.default('Table identifier missing', { exit: 0, errorId: 'IMP-1' });
    }
    if (!importConfig.file_id) {
        if (!uploadFile) {
            throw new error_1.default('No CSV file provided', { exit: 0, errorId: 'IMP-2' });
        }
        const fileAPI = yield (0, endpoints_1.filestoreAPI)({ env });
        const allFolders = (yield fileAPI.getAllFolders());
        if (allFolders.length > 0) {
            const choices = allFolders.map((folderDetail) => prompt_1.default.choice(folderDetail.folder_name, {
                value: folderDetail
            }));
            const folderAns = yield prompt_1.default.ask(prompt_1.default.question('folder', 'Select a folder to which you want to upload the file to : ', {
                type: 'list',
                choices,
                when: !importConfig.folder_id
            }));
            folderAns.folder =
                folderAns.folder === undefined
                    ? allFolders.find((folderDetails) => folderDetails.folder_name === importConfig.folder_id ||
                        folderDetails.id === importConfig.folder_id)
                    : folderAns.folder;
            if (!folderAns.folder) {
                throw new error_1.default('Folder not present', {
                    exit: 1,
                    errorId: 'IMP-4',
                    arg: [
                        ansi_colors_1.bold.red(importConfig.folder_id || ''),
                        (0, ansi_colors_1.italic)(configPath || ''),
                        (0, ansi_colors_1.bold)(uploadFile),
                        (0, ansi_colors_1.bold)('Unique ID'),
                        (0, ansi_colors_1.italic)('file_id')
                    ]
                });
            }
            const fileDetails = (yield fileAPI.uploadFile(folderAns.folder.id, fs_1.SYNC.getReadStream((0, path_1.resolve)(uploadFile))));
            importConfig.file_id = fileDetails.id;
        }
        else {
            throw new error_1.default('folder needed', { exit: 0, errorId: 'IMP-3' });
        }
    }
    const bulkAPI = yield (0, endpoints_1.bulkDSAPI)({ env });
    const writeResponse = (yield bulkAPI.write(tableId, importConfig));
    (0, logger_1.info)();
    (0, logger_1.success)(`Successfully scheduled import job for table "${tableId}" with jobid "${writeResponse.job_id}"`);
    if ((0, env_1.isPrimaryShell)()) {
        const exitListeners = process.listeners('exit');
        process.removeAllListeners('exit');
        const allGlobalOpts = runtime_store_1.default.get('opts.globalOpts', {});
        const optsArr = Object.keys(allGlobalOpts).reduce((arr, key) => {
            arr.push('--' + key);
            if (allGlobalOpts[key]) {
                arr.push(allGlobalOpts[key]);
            }
            return arr;
        }, (env === 'Production' ? ['--production'] : []));
        (0, shell_1.spawn)('catalyst', ['ds:status', 'import', writeResponse.job_id + '', ...optsArr], {
            shell: true,
            stdio: 'inherit'
        }).SYNC();
        (exitListeners || []).forEach((listner) => {
            process.addListener('exit', listner);
        });
    }
}));
