// zcatalyst-cli/lib/event_generate/webapp.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const moment_1 = __importDefault(require("moment"));
const endpoints_1 = require("../endpoints");
const runtime_store_1 = __importDefault(require("../runtime-store"));
const constants_1 = require("../util_modules/constants");
const js_1 = require("../util_modules/js");
const project_1 = require("../util_modules/project");
exports.default = (eventJson) => __awaiter(void 0, void 0, void 0, function* () {
    const webappDetails = (yield (yield (0, endpoints_1.clientAPI)()).getWebappDetails());
    const selectedWebapp = webappDetails !== undefined && Object.keys(webappDetails).length > 0
        ? webappDetails
        : {
            app_name: js_1.JS.randomString(js_1.JS.random(5, 10)),
            app_id: js_1.JS.randomNumber(15),
            url_prefix: (0, project_1.getDomainPrefix)()
        };
    const userDetails = {
        email_id: runtime_store_1.default.get('user.Email'),
        user_id: js_1.JS.randomNumber(15),
        is_confirmed: js_1.JS.random(0, 1) === 1 ? true : false,
        last_name: js_1.JS.randomString(5),
        first_name: js_1.JS.randomString(10),
        zuid: js_1.JS.randomNumber(10)
    };
    const data = {
        uploaded_time: (0, moment_1.default)().format('MMM DD, YYYY hh:mm A'),
        project_details: {
            id: (0, project_1.getProjectId)(),
            project_name: (0, project_1.getProjectName)()
        },
        app_name: selectedWebapp.app_name,
        uploaded_by: userDetails,
        app_id: selectedWebapp.app_id,
        url_prefix: selectedWebapp.url_prefix
    };
    eventJson = eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.source_id, 'g'), 'NULL');
    eventJson = eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.rule_id, 'g'), 'NULL');
    return eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.data, 'g'), JSON.stringify(data));
});
