// zcatalyst-angular-schematics/collectionTemplate.js
module.exports = {
	schematics: {
		'ng-new': {
			factory: '{{location}}/@schematics/angular/ng-new',
			schema: '{{location}}/@schematics/angular/ng-new/schema.json',
			description: 'Create an Angular workspace.',
			hidden: true
		},
		workspace: {
			factory: '{{location}}/@schematics/angular/workspace',
			schema: '{{location}}/@schematics/angular/workspace/schema.json',
			description: 'Create an Angular workspace.',
			hidden: true
		},
		'service-worker': {
			aliases: ['serviceWorker'],
			factory: '{{location}}/@schematics/angular/service-worker',
			description: 'Initializes a service worker setup.',
			schema: '{{location}}/@schematics/angular/service-worker/schema.json'
		},
		application: {
			aliases: ['app'],
			factory: './lib/application',
			schema: '{{location}}/@schematics/angular/application/schema.json',
			description: 'Create an Angular application.'
		},
		e2e: {
			factory: '{{location}}/@schematics/angular/e2e',
			schema: '{{location}}/@schematics/angular/e2e/schema.json',
			description: 'Create an Angular e2e application.',
			hidden: true
		},
		class: {
			aliases: ['cl'],
			factory: '{{location}}/@schematics/angular/class',
			description: 'Create a class.',
			schema: '{{location}}/@schematics/angular/class/schema.json'
		},
		component: {
			aliases: ['c'],
			factory: '{{location}}/@schematics/angular/component',
			description: 'Create an Angular component.',
			schema: '{{location}}/@schematics/angular/component/schema.json'
		},
		directive: {
			aliases: ['d'],
			factory: '{{location}}/@schematics/angular/directive',
			description: 'Create an Angular directive.',
			schema: '{{location}}/@schematics/angular/directive/schema.json'
		},
		enum: {
			aliases: ['e'],
			factory: '{{location}}/@schematics/angular/enum',
			description: 'Create an enumeration.',
			schema: '{{location}}/@schematics/angular/enum/schema.json'
		},
		guard: {
			aliases: ['g'],
			factory: '{{location}}/@schematics/angular/guard',
			description: 'Create a guard.',
			schema: '{{location}}/@schematics/angular/guard/schema.json'
		},
		resolver: {
			aliases: ['r'],
			factory: '{{location}}/@schematics/angular/resolver',
			description: 'Create a resolver.',
			schema: '{{location}}/@schematics/angular/resolver/schema.json'
		},
		interceptor: {
			factory: '{{location}}/@schematics/angular/interceptor',
			description: 'Create an interceptor.',
			schema: '{{location}}/@schematics/angular/interceptor/schema.json'
		},
		interface: {
			aliases: ['i'],
			factory: '{{location}}/@schematics/angular/interface',
			description: 'Create an interface.',
			schema: '{{location}}/@schematics/angular/interface/schema.json'
		},
		module: {
			aliases: ['m'],
			factory: '{{location}}/@schematics/angular/module',
			description: 'Create an Angular module.',
			schema: '{{location}}/@schematics/angular/module/schema.json'
		},
		pipe: {
			aliases: ['p'],
			factory: '{{location}}/@schematics/angular/pipe',
			description: 'Create an Angular pipe.',
			schema: '{{location}}/@schematics/angular/pipe/schema.json'
		},
		service: {
			aliases: ['s'],
			factory: '{{location}}/@schematics/angular/service',
			description: 'Create an Angular service.',
			schema: '{{location}}/@schematics/angular/service/schema.json'
		},
		universal: {
			factory: '{{location}}/@schematics/angular/universal',
			description: 'Create an Angular universal app.',
			schema: '{{location}}/@schematics/angular/universal/schema.json',
			hidden: true
		},
		'app-shell': {
			aliases: ['appShell'],
			factory: '{{location}}/@schematics/angular/app-shell',
			description: 'Create an app shell.',
			schema: '{{location}}/@schematics/angular/app-shell/schema.json'
		},
		library: {
			aliases: ['lib'],
			factory: '{{location}}/@schematics/angular/library',
			schema: '{{location}}/@schematics/angular/library/schema.json',
			description: 'Generate a library project for Angular.'
		},
		'web-worker': {
			aliases: ['webWorker'],
			factory: '{{location}}/@schematics/angular/web-worker',
			schema: '{{location}}/@schematics/angular/web-worker/schema.json',
			description: 'Create a Web Worker.'
		}
	}
};
