// zcatalyst-cli/lib/util_modules/constants/index.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENVPATH = exports.RUNTIME = exports.IAC = exports.PLUGIN = exports.ORIGIN = exports.TEMPLATE = exports.SCOPE = exports.REMOTE_REF = exports.REGEX = exports.REFERENCE = exports.PLACEHOLDER = exports.INTEG = exports.FOLDERNAME = exports.FN_TYPE = exports.DC_TYPE = exports.FILENAME = exports.EVENT_SOURCE = exports.EVENT_REF = exports.DEFAULT = exports.CLIQ = exports.AUTH = exports.APIG_RULES = void 0;
var apig_rules_1 = require("./lib/apig-rules");
Object.defineProperty(exports, "APIG_RULES", { enumerable: true, get: function () { return __importDefault(apig_rules_1).default; } });
var auth_1 = require("./lib/auth");
Object.defineProperty(exports, "AUTH", { enumerable: true, get: function () { return __importDefault(auth_1).default; } });
var cliq_handlers_1 = require("./lib/cliq-handlers");
Object.defineProperty(exports, "CLIQ", { enumerable: true, get: function () { return __importDefault(cliq_handlers_1).default; } });
var default_1 = require("./lib/default");
Object.defineProperty(exports, "DEFAULT", { enumerable: true, get: function () { return __importDefault(default_1).default; } });
var event_action_1 = require("./lib/event-action");
Object.defineProperty(exports, "EVENT_REF", { enumerable: true, get: function () { return __importDefault(event_action_1).default; } });
var event_source_1 = require("./lib/event-source");
Object.defineProperty(exports, "EVENT_SOURCE", { enumerable: true, get: function () { return __importDefault(event_source_1).default; } });
var file_names_1 = require("./lib/file-names");
Object.defineProperty(exports, "FILENAME", { enumerable: true, get: function () { return __importDefault(file_names_1).default; } });
var dc_type_1 = require("./lib/dc-type");
Object.defineProperty(exports, "DC_TYPE", { enumerable: true, get: function () { return __importDefault(dc_type_1).default; } });
var fn_type_1 = require("./lib/fn-type");
Object.defineProperty(exports, "FN_TYPE", { enumerable: true, get: function () { return __importDefault(fn_type_1).default; } });
var folder_names_1 = require("./lib/folder-names");
Object.defineProperty(exports, "FOLDERNAME", { enumerable: true, get: function () { return __importDefault(folder_names_1).default; } });
var integ_1 = require("./lib/integ");
Object.defineProperty(exports, "INTEG", { enumerable: true, get: function () { return __importDefault(integ_1).default; } });
var placeholders_1 = require("./lib/placeholders");
Object.defineProperty(exports, "PLACEHOLDER", { enumerable: true, get: function () { return __importDefault(placeholders_1).default; } });
var ref_mapping_1 = require("./lib/ref-mapping");
Object.defineProperty(exports, "REFERENCE", { enumerable: true, get: function () { return __importDefault(ref_mapping_1).default; } });
var regex_1 = require("./lib/regex");
Object.defineProperty(exports, "REGEX", { enumerable: true, get: function () { return __importDefault(regex_1).default; } });
var remote_mapping_1 = require("./lib/remote-mapping");
Object.defineProperty(exports, "REMOTE_REF", { enumerable: true, get: function () { return __importDefault(remote_mapping_1).default; } });
var scopes_1 = require("./lib/scopes");
Object.defineProperty(exports, "SCOPE", { enumerable: true, get: function () { return __importDefault(scopes_1).default; } });
var template_1 = require("./lib/template");
Object.defineProperty(exports, "TEMPLATE", { enumerable: true, get: function () { return __importDefault(template_1).default; } });
var urls_1 = require("./lib/urls");
Object.defineProperty(exports, "ORIGIN", { enumerable: true, get: function () { return __importDefault(urls_1).default; } });
var plugin_1 = require("./lib/plugin");
Object.defineProperty(exports, "PLUGIN", { enumerable: true, get: function () { return __importDefault(plugin_1).default; } });
var iac_1 = require("./lib/iac");
Object.defineProperty(exports, "IAC", { enumerable: true, get: function () { return __importDefault(iac_1).default; } });
var runtime_1 = require("./lib/runtime");
Object.defineProperty(exports, "RUNTIME", { enumerable: true, get: function () { return __importDefault(runtime_1).default; } });
var env_path_1 = require("./lib/env-path");
Object.defineProperty(exports, "ENVPATH", { enumerable: true, get: function () { return __importDefault(env_path_1).default; } });
