// zcatalyst-cli/lib/endpoints/lib/appsail.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const constants_1 = require("../../util_modules/constants");
const logger_1 = require("../../util_modules/logger");
class AppSail {
    constructor(projectId, opts) {
        this.opts = opts;
        this.projectId = projectId;
    }
    deploy(sourceFsStream, { stack, name, contentLength, command, memory, platform, envVariables = {} }) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = {
                stack,
                name,
                code: sourceFsStream
            };
            if (typeof memory === 'number') {
                formData.memory = memory;
            }
            if (typeof platform === 'string') {
                formData.platform = platform;
            }
            const configuration = {};
            if (typeof command === 'string') {
                configuration.startup_command = command;
            }
            if (typeof envVariables === 'object' && Object.keys(envVariables).length > 0) {
                configuration.environment = {
                    variables: envVariables
                };
            }
            if (Object.keys(configuration).length > 0) {
                formData.configuration = JSON.stringify(configuration);
            }
            const res = yield new api_1.default(this.opts).put(`/baas/v1/project/${this.projectId}/appsail/upsert`, {
                formData,
                json: false,
                log: {
                    uploadProgress: {
                        title: `AppSail[${name}]`,
                        total: contentLength,
                        progressType: 'upload'
                    },
                    stream: sourceFsStream,
                    awaitingSpinner: `deploying Appsail[${name}]`
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('save AppSail response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    updateConfig(appSailId, env) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/baas/v1/project/${this.projectId}/appsail/${appSailId}/configuration`, {
                body: {
                    environment: {
                        variables: env
                    }
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('Update configuration response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    downloadJetty() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default({
                origin: constants_1.ORIGIN.zohoCDN,
                authNeeded: false,
                json: false,
                isExternal: true
            }).get('/catalyst/appsail/1.0.0/jetty_runtime_dependency.zip', { encoding: null });
            if (res.body) {
                return res.body;
            }
            (0, logger_1.debug)('Download JETTY server response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = AppSail;
