// zcatalyst-cli/lib/endpoints/lib/code-deck.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _CodeDeck_instances, _CodeDeck_manifest;
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const constants_1 = require("../../util_modules/constants");
const logger_1 = require("../../util_modules/logger");
class CodeDeck {
    constructor(opts) {
        _CodeDeck_instances.add(this);
        this.opts = opts;
    }
    data() {
        return __awaiter(this, void 0, void 0, function* () {
            const manifestData = yield __classPrivateFieldGet(this, _CodeDeck_instances, "m", _CodeDeck_manifest).call(this);
            if (!manifestData['data.json']) {
                throw new error_1.default('Invalid manifest data', { exit: 2 });
            }
            const res = yield new api_1.default(this.opts).get(`/catalyst-library/codelib/${manifestData['data.json']}`, {
                json: true,
                origin: constants_1.ORIGIN.zohoCDN,
                headers: {
                    'Cache-Control': 'no-cache'
                }
            });
            if (res.body) {
                return res.body;
            }
            (0, logger_1.debug)('Code Lib data json response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = CodeDeck;
_CodeDeck_instances = new WeakSet(), _CodeDeck_manifest = function _CodeDeck_manifest() {
    return __awaiter(this, void 0, void 0, function* () {
        const res = yield new api_1.default(this.opts).get('/catalyst-library/codelib/config.json', {
            json: true,
            origin: constants_1.ORIGIN.zohoCDN,
            headers: {
                'Cache-Control': 'no-cache'
            }
        });
        if (res.body) {
            return res.body;
        }
        (0, logger_1.debug)('Code Lib manifest json response from server : ' + res.body);
        throw new error_1.default('Server Error: Unexpected Response from server.', {
            exit: 2
        });
    });
};
