// zcatalyst-cli/lib/endpoints/lib/functions.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const constants_1 = require("../../util_modules/constants");
const logger_1 = require("../../util_modules/logger");
class Functions {
    constructor(projectId, opts) {
        this.opts = opts;
        this.projectId = projectId;
    }
    getAllFunctions(type) {
        return __awaiter(this, void 0, void 0, function* () {
            if (type === undefined) {
                type = Object.values(constants_1.REMOTE_REF.functions.type)
                    .filter((t) => t !== constants_1.REMOTE_REF.functions.type[constants_1.FN_TYPE.advanced])
                    .join(',');
            }
            const opts = {
                qs: {
                    type
                }
            };
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/function`, opts);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get all function response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getFunction(fnId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/function/${fnId}`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('Get function response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    updateConfig(functionId, env) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/baas/v1/project/${this.projectId}/function/${functionId}/configuration`, {
                body: {
                    environment: {
                        variables: env
                    }
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('Update configuration response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    deploy(sourceFsStream, { stack, name, type, memory, envVariables, contentLength }) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = {
                stack,
                name,
                type
            };
            const configuration = {};
            if (sourceFsStream !== undefined) {
                formData.code = sourceFsStream;
            }
            if (memory !== undefined) {
                configuration.memory = memory;
            }
            if (envVariables !== undefined) {
                configuration.environment = {
                    variables: envVariables
                };
            }
            if (Object.keys(configuration).length > 0) {
                formData.configuration = JSON.stringify(configuration);
            }
            const log = contentLength
                ? {
                    uploadProgress: {
                        title: `[${name}]`,
                        total: contentLength,
                        progressType: 'upload'
                    },
                    stream: sourceFsStream,
                    awaitingSpinner: `deploying function [${name}]`
                }
                : undefined;
            const res = yield new api_1.default(this.opts).put(`/baas/v1/project/${this.projectId}/function/upsert`, {
                formData,
                json: false,
                log
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('save function response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    download(functionId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/function/${functionId}/download`, {
                json: false,
                encoding: null
            });
            if (res.body) {
                return res.body;
            }
            (0, logger_1.debug)('download function response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    delete(functionId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).delete(`/baas/v1/project/${this.projectId}/function/${functionId}`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get all function response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = Functions;
