// zcatalyst-cli/lib/endpoints/lib/project.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const logger_1 = require("../../util_modules/logger");
class ProjectAPI {
    constructor(opts, org) {
        this.opts = opts;
        this.opts.envId = org;
    }
    getProject(projectId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get('/baas/v1/project/' + projectId);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get project response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getAllProjects() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get('/baas/v1/project');
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get all project response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    createProject(name) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post('/baas/v1/project', {
                body: {
                    project_name: name
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('create project response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getCurrentProjectUser(projectId, headers = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(Object.assign(Object.assign({}, this.opts), { headers, authNeeded: false, resolveOnError: true })).get(`/baas/v1/project/${projectId}/project-user/current`);
            if (res.body && res.body.data) {
                if (typeof res.body.data.email_id === 'undefined') {
                    return null;
                }
                return res.body.data;
            }
            (0, logger_1.debug)('get project response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = ProjectAPI;
