// zcatalyst-cli/lib/endpoints/lib/queue.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const logger_1 = require("../../util_modules/logger");
class Queue {
    constructor(projectId, opts) {
        this.opts = opts;
        this.projectId = projectId;
    }
    getAllTopics() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/queue`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get all topics response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    produce(data, topicId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/baas/v1/project/${this.projectId}/queue/${topicId}`, {
                body: {
                    value: data
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('produce data to queue response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    consume(topicId) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/queue/${topicId}/records`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            else if (res.status === 204) {
                return undefined;
            }
            (0, logger_1.debug)('consume data from queue response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = Queue;
