// zcatalyst-cli/lib/util_modules/constants/lib/apig-rules.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Object.freeze([
    {
        name: 'catalyst_default_cdn',
        source_endpoint: '/__catalyst/{path1:(.*)}',
        target_endpoint: '/__catalyst/{path1}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_signin_redirect',
        source_endpoint: '/baas/{projectId:[a-zA-Z0-9_]+}/signin-redirect',
        target_endpoint: '/baas/{projectId}/signin-redirect',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_check_auth',
        source_endpoint: '/baas/{projectId:[a-zA-Z0-9_]+}/check-auth',
        target_endpoint: '/baas/{projectId}/check-auth',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_zcql',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/query',
        target_endpoint: '/baas/v1/project/{projectId}/query',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_search',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/search',
        target_endpoint: '/baas/v1/project/{projectId}/search',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_notification',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/project-user/notification-client',
        target_endpoint: '/baas/v1/project/{projectId}/project-user/notification-client',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_all_Table',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table',
        target_endpoint: '/baas/v1/project/{projectId}/table',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_table',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_all_columns',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/column',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/column',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_column',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/column/{columnId:[A-Za-z0-9_]+}',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/column/{columnId}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_all_rows',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_row',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row/{rowId:[0-9]+}',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row/{rowId}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_add_rows',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_update_rows',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row',
        method: 'PUT',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_delete_row',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row/{rowId:[0-9]+}',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row/{rowId}',
        method: 'DELETE',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_delete_rows',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/table/{tableId:[A-Za-z0-9_]+}/row',
        target_endpoint: '/baas/v1/project/{projectId}/table/{tableId}/row',
        method: 'DELETE',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_signup',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/project-user/signup',
        target_endpoint: '/baas/v1/project/{projectId}/project-user/signup',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_current_user',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/project-user/current',
        target_endpoint: '/baas/v1/project/{projectId}/project-user/current',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_forget_password',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/project-user/forgotpassword',
        target_endpoint: '/baas/v1/project/{projectId}/project-user/forgotpassword',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_all_folders',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder',
        target_endpoint: '/baas/v1/project/{projectId}/folder',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_folder',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder/{folderId:([A-Za-z0-9_]+)}',
        target_endpoint: '/baas/v1/project/{projectId}/folder/{folderId}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_upload_file',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder/{folderId:[A-Za-z0-9_]+}/file',
        target_endpoint: '/baas/v1/project/{projectId}/folder/{folderId}/file',
        method: 'POST',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_get_file',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder/{folderId:[A-Za-z0-9_]+}/file/{fileId:[0-9]+}',
        target_endpoint: '/baas/v1/project/{projectId}/folder/{folderId}/file/{fileId}',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_delete_file',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder/{folderId:[A-Za-z0-9_]+}/file/{fileId:[0-9]+}',
        target_endpoint: '/baas/v1/project/{projectId}/folder/{folderId}/file/{fileId}',
        method: 'DELETE',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_download_file',
        source_endpoint: '/baas/v1/project/{projectId:[0-9]+}/folder/{folderId:[A-Za-z0-9_]+}/file/{fileId:[0-9]+}/download',
        target_endpoint: '/baas/v1/project/{projectId}/folder/{folderId}/file/{fileId}/download',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_iam_oauthorize',
        source_endpoint: '/oauthorize',
        target_endpoint: '/oauthorize',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_confirm_account',
        source_endpoint: '/baas/confirmaccount',
        target_endpoint: '/baas/confirmaccount',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_reset_password',
        source_endpoint: '/baas/resetpassword',
        target_endpoint: '/baas/resetpassword',
        method: 'GET',
        authentication: undefined,
        status: true
    },
    {
        name: 'catalyst_default_logout',
        source_endpoint: '/baas/logout',
        target_endpoint: '/baas/logout',
        method: 'GET',
        authentication: undefined,
        status: true
    }
]);
