// zcatalyst-cli/lib/util_modules/constants/lib/cliq-handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const file_names_1 = __importDefault(require("./file-names"));
const { node_main, node_package, java_main, java_classpath, java_project, python_main, python_requirements } = file_names_1.default.functions;
const template_root = (0, path_1.join)(__dirname, '../../../../templates');
const handlers = {
    bot_handler: 'bot_handler',
    command_handler: 'command_handler',
    messageaction_handler: 'messageaction_handler',
    widget_handler: 'widget_handler',
    function_handler: 'function_handler',
    installation_handler: 'installation_handler',
    installation_validator: 'installation_validator'
};
exports.default = Object.freeze({
    event_template_dir: (0, path_1.join)(template_root, './event/integ/cliq/'),
    node_handlers_template_dir: './handlers',
    java_handlers_template_dir: './com/handlers',
    python_handlers_template_dir: './handlers',
    node_handlers_path: 'handlers/',
    java_handlers_path: 'com.handlers.',
    python_handlers_path: 'handlers/',
    node_template_contents_initial_copy: [node_main, node_package, file_names_1.default.catalyst_config],
    java_template_contents_initial_copy: [
        java_main,
        java_classpath,
        java_project,
        file_names_1.default.catalyst_config
    ],
    python_template_contents_initial_copy: [
        python_main,
        file_names_1.default.catalyst_config,
        python_requirements
    ],
    component_handlers: Object.keys(handlers),
    node_handler_file_mapping: {
        [handlers.bot_handler]: 'bot-handler.js',
        [handlers.command_handler]: 'command-handler.js',
        [handlers.messageaction_handler]: 'message-action-handler.js',
        [handlers.widget_handler]: 'widget-handler.js',
        [handlers.function_handler]: 'function-handler.js',
        [handlers.installation_handler]: 'installation-handler.js',
        [handlers.installation_validator]: 'installation-validator.js'
    },
    java_handler_file_mapping: {
        [handlers.bot_handler]: 'BotHandler.java',
        [handlers.command_handler]: 'CommandHandler.java',
        [handlers.messageaction_handler]: 'MessageActionHandler.java',
        [handlers.widget_handler]: 'WidgetHandler.java',
        [handlers.function_handler]: 'FunctionHandler.java',
        [handlers.installation_handler]: 'InstallationHandler.java',
        [handlers.installation_validator]: 'InstallationValidator.java'
    },
    python_handler_file_mapping: {
        [handlers.bot_handler]: 'bot_handler.py',
        [handlers.command_handler]: 'command_handler.py',
        [handlers.messageaction_handler]: 'message_action_handler.py',
        [handlers.widget_handler]: 'widget_handler.py',
        [handlers.function_handler]: 'function_handler.py',
        [handlers.installation_handler]: 'installation_handler.py',
        [handlers.installation_validator]: 'installation_validator.py'
    },
    handlers_event_template: {
        ['bot']: [
            'welcome_handler',
            'message_handler',
            'context_handler',
            'mention_handler',
            'action_handler',
            'incomingwebhook_handler',
            'participation_handler'
        ],
        ['command']: ['execution_handler', 'suggestion_handler'],
        ['messageaction']: ['execution_handler'],
        ['widget']: ['view_handler'],
        ['function']: [
            'button_handler',
            'form_handler',
            'form_change_handler',
            'form_values_handler',
            'widget_button_handler'
        ],
        ['installation']: ['installation_handler', 'installation_validator']
    }
});
