// zcatalyst-cli/lib/util_modules/constants/lib/placeholders.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Object.freeze({
    client: {
        port: '{{_PORT_}}',
        package: {
            name: '{{_NAME_}}'
        }
    },
    functions: {
        node_package: {
            name: '{{_NAME_}}',
            main: '{{_MAIN_}}',
            author: '{{_AUTHOR_}}',
            dependencies: '{{_DEPENDENCIES_}}'
        },
        java_class: '{{_CLASS_}}',
        java_name: '{{_NAME_}}',
        python_package: {
            name: '{{_NAME_}}',
            main: '{{_MAIN_}}',
            sdkVersion: '{{_VERSION_}}'
        }
    },
    catalyst_config: {
        type: '{{_TYPE_}}',
        stack: '{{_STACK_}}',
        integ_config: '{{__INTEG_CONFIG__}}'
    },
    event_data: {
        project_id: '{{__PROJECT_ID__}}',
        project_name: '{{__PROJECT_NAME__}}',
        eventbus_id: '{{__EVENTBUS_ID__}}',
        eventbus_name: '{{__EVENTBUS_NAME__}}',
        action: '{{__ACTION__}}',
        source: '{{__SOURCE__}}',
        source_id: '{{__SOURCE_ID__}}',
        event_time: '{{__EVENT_TIME__}}',
        data: '{{__DATA__}}',
        rule_id: '{{__RULE_ID__}}'
    }
});
