// zcatalyst-cli/lib/util_modules/constants/lib/scopes.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Object.freeze({
    api_read: 'ZohoCatalyst.apigateway.apis.READ',
    api_create: 'ZohoCatalyst.apigateway.apis.CREATE',
    api_update: 'ZohoCatalyst.apigateway.apis.UPDATE',
    api_delete: 'ZohoCatalyst.apigateway.apis.DELETE',
    apig_read: 'ZohoCatalyst.apigateway.READ',
    apig_update: 'ZohoCatalyst.apigateway.UPDATE',
    segment: 'ZohoCatalyst.segments.ALL',
    cache_read: 'ZohoCatalyst.cache.READ',
    cache_create: 'ZohoCatalyst.cache.CREATE',
    cache_delete: 'ZohoCatalyst.cache.DELETE',
    queue: 'ZohoCatalyst.queue.ALL',
    queue_data_read: 'ZohoCatalyst.queue.data.READ',
    queue_data_create: 'ZohoCatalyst.queue.data.CREATE',
    projects: 'ZohoCatalyst.projects.ALL',
    project_export_read: 'ZohoCatalyst.project.export.READ',
    project_export_create: 'ZohoCatalyst.project.export.CREATE',
    project_import_read: 'ZohoCatalyst.project.import.READ',
    project_import_create: 'ZohoCatalyst.project.import.CREATE',
    project_import_delete: 'ZohoCatalyst.project.import.DELETE',
    cron: 'ZohoCatalyst.cron.ALL',
    zcql: 'ZohoCatalyst.zcql.CREATE',
    webapp: 'ZohoCatalyst.webapp.ALL',
    project_user_read: 'ZohoCatalyst.projects.users.READ',
    project_user_create: 'ZohoCatalyst.projects.users.CREATE',
    project_user_delete: 'ZohoCatalyst.projects.users.DELETE',
    project_user_update: 'ZohoCatalyst.projects.users.UPDATE',
    email_send: 'ZohoCatalyst.email.CREATE',
    datastore: 'ZohoCatalyst.tables.ALL',
    row: 'ZohoCatalyst.tables.rows.ALL',
    column: 'ZohoCatalyst.tables.columns.ALL',
    bulk_read: 'ZohoCatalyst.tables.bulk.READ',
    bulk_create: 'ZohoCatalyst.tables.bulk.CREATE',
    folders: 'ZohoCatalyst.folders.ALL',
    file_create: 'ZohoCatalyst.files.CREATE',
    file_read: 'ZohoCatalyst.files.READ',
    file_delete: 'ZohoCatalyst.files.DELETE',
    functions: 'ZohoCatalyst.functions.ALL',
    functions_execution: 'ZohoCatalyst.functions.EXECUTE',
    event_listeners: 'ZohoCatalyst.eventlistners.ALL',
    event_produce: 'ZohoCatalyst.eventlisteners.produce',
    security_read: 'ZohoCatalyst.security.rules.Read',
    security_update: 'ZohoCatalyst.security.rules.Update',
    notification_mobile: 'ZohoCatalyst.notifications.mobile',
    notification_web: 'ZohoCatalyst.notifications.web',
    search_read: 'ZohoCatalyst.search.READ',
    zia: 'ZohoCatalyst.mlkit.READ',
    quick_ml: 'QuickML.deployment.READ',
    pdf_shot: 'ZohoCatalyst.pdfshot.execute',
    dataverse: 'ZohoCatalyst.dataverse.execute',
    circuits_execute: 'ZohoCatalyst.circuits.execute',
    circuits_exe_read: 'ZohoCatalyst.circuits.execution.READ',
    circuits_exe_delete: 'ZohoCatalyst.circuits.execution.DELETE',
    cli_logs: 'ZohoCatalyst.cli.logs.CREATE',
    appsail_read: 'ZohoCatalyst.appsail.Read',
    appsail_create: 'ZohoCatalyst.appsail.Create',
    appsail_update: 'ZohoCatalyst.appsail.Update',
    appsail_delete: 'ZohoCatalyst.appsail.Delete',
    iam_profile: 'AaaServer.profile.READ',
    stratus_create: 'Stratus.sdkop.CREATE',
    stratus_read: 'Stratus.sdkop.READ'
});
