// zcatalyst-cli/lib/util_modules/constants/lib/template.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const constants_1 = require("../../constants");
const TEMPLATE_ROOT = (0, path_1.resolve)(__dirname, '../../../../templates');
const FUNCTIONS_NODE_ROOT = (0, path_1.join)(TEMPLATE_ROOT, 'init', 'functions', 'node');
const FUNCTIONS_JAVA_ROOT = (0, path_1.join)(TEMPLATE_ROOT, 'init', 'functions', 'java');
const FUNCTIONS_PYTHON_ROOT = (0, path_1.join)(TEMPLATE_ROOT, 'init', 'functions', 'python');
exports.default = Object.freeze({
    banner: (0, path_1.join)(TEMPLATE_ROOT, 'banner.txt'),
    login_success: (0, path_1.join)(TEMPLATE_ROOT, 'loginSuccess.html'),
    login_fail: (0, path_1.join)(TEMPLATE_ROOT, 'loginFailure.html'),
    client: {
        basic: {
            new: (0, path_1.join)(TEMPLATE_ROOT, 'init', 'client', 'basic'),
            socket: (0, path_1.join)(TEMPLATE_ROOT, 'web-socket.txt')
        },
        lyte: (0, path_1.join)(TEMPLATE_ROOT, 'init', 'client', 'lyte'),
        react: {
            js: (0, path_1.join)(TEMPLATE_ROOT, 'init', 'client', 'react', 'react_js'),
            ts: (0, path_1.join)(TEMPLATE_ROOT, 'init', 'client', 'react', 'react_ts')
        }
    },
    applogic: {
        node: (0, path_1.join)(TEMPLATE_ROOT, 'init', 'applogic', 'node')
    },
    functions: {
        node: {
            [constants_1.FN_TYPE.basic]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'bio'),
            [constants_1.FN_TYPE.event]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'event'),
            [constants_1.FN_TYPE.cron]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'cron'),
            [constants_1.FN_TYPE.advanced]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'aio'),
            [constants_1.FN_TYPE.integration]: {
                [constants_1.INTEG.services.cliq]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'integ', 'cliq'),
                [constants_1.INTEG.services.convokraft]: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'integ', 'convokraft')
            },
            [constants_1.FN_TYPE.browserLogic]: {
                Playwright: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'browserlogic', 'playwright'),
                Puppeteer: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'browserlogic', 'puppeteer'),
                Selenium: (0, path_1.join)(FUNCTIONS_NODE_ROOT, 'browserlogic', 'selenium')
            }
        },
        java: {
            [constants_1.FN_TYPE.basic]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'bio'),
            [constants_1.FN_TYPE.event]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'event'),
            [constants_1.FN_TYPE.cron]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'cron'),
            [constants_1.FN_TYPE.advanced]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'aio'),
            [constants_1.FN_TYPE.integration]: {
                [constants_1.INTEG.services.cliq]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'integ', 'cliq'),
                [constants_1.INTEG.services.convokraft]: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'integ', 'convokraft')
            },
            [constants_1.FN_TYPE.browserLogic]: {
                Selenium: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'browserlogic', 'selenium'),
                Playwright: (0, path_1.join)(FUNCTIONS_JAVA_ROOT, 'browserlogic', 'playwright')
            }
        },
        python: {
            [constants_1.FN_TYPE.basic]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'bio'),
            [constants_1.FN_TYPE.event]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'event'),
            [constants_1.FN_TYPE.cron]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'cron'),
            [constants_1.FN_TYPE.advanced]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'aio'),
            [constants_1.FN_TYPE.integration]: {
                [constants_1.INTEG.services.cliq]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'integ', 'cliq'),
                [constants_1.INTEG.services.convokraft]: (0, path_1.join)(FUNCTIONS_PYTHON_ROOT, 'integ', 'convokraft')
            }
        }
    },
    event_data: (0, path_1.join)(TEMPLATE_ROOT, 'event', 'data.json')
});
