// zcatalyst-cli/lib/util_modules/constants/lib/urls.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const env_js_1 = require("../../env.js");
const dc_js_1 = require("../../../dc.js");
const dc_type_js_1 = __importDefault(require("./dc-type.js"));
class URL {
    static get auth() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._auth;
        }
        if (dc.value === dc_type_js_1.default.ca.value) {
            return URL._auth.replace('.zoho.com', '.zohocloud.ca');
        }
        return URL._auth.replace('.com', dc.ext);
    }
    static get admin() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._admin;
        }
        if (dc.value === dc_type_js_1.default.ca.value) {
            return URL._admin.replace('.zoho.com', '.zohocloud.ca');
        }
        return URL._admin.replace('.com', dc.ext);
    }
    static get app() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._app;
        }
        return URL._app.replace('.com', dc.ext);
    }
    static get catalystStatic() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._zohoStatic;
        }
        if (dc.value === dc_type_js_1.default.ca.value) {
            return URL._zohoStatic.replace('.zoho.com', '.zohocloud.ca');
        }
        return URL._zohoStatic.replace(dc_type_js_1.default.us.ext, dc.ext);
    }
    static get console() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._console;
        }
        if (dc.value === dc_type_js_1.default.ca.value) {
            return URL._console.replace('.zoho.com', '.zohocloud.ca');
        }
        return URL._console.replace('.com', dc.ext);
    }
}
exports.default = URL;
URL._auth = (0, env_js_1.envOverride)('CATALYST_AUTH_URL', 'https://accounts.zoho.com');
URL._admin = (0, env_js_1.envOverride)('CATALYST_ADMIN_URL', 'https://api.catalyst.zoho.com');
URL._app = (0, env_js_1.envOverride)('CATALYST_APP_URL', 'https://catalystserverless.com');
URL._zohoStatic = (0, env_js_1.envOverride)('ZOHO_STATIC', 'https://www.zoho.com/catalyst');
URL._console = (0, env_js_1.envOverride)('CATALYST_CONSOLE_URL', 'https://console.catalyst.zoho.com');
URL.zohoCDN = (0, env_js_1.envOverride)('ZOHO_CDN', 'https://static.zohocdn.com');
URL.external = Object.freeze({
    gitHubAPI: 'https://api.github.com',
    gitHubWeb: 'https://github.com'
});
