// zcatalyst-cli/lib/util_modules/config/lib/appSail.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllTargetDetails = exports.getTargetDetails = exports.raw = void 0;
const error_1 = __importDefault(require("../../../error"));
const runtime_store_1 = __importDefault(require("../../../runtime-store"));
const constants_1 = require("../../constants");
const fs_1 = require("../../fs");
const logger_1 = require("../../logger");
const project_1 = require("../../project");
function raw(throwError = false) {
    const config = runtime_store_1.default.get('config', null);
    if (config === null) {
        if (throwError) {
            throw new error_1.default(constants_1.FILENAME.config + ' file is required', { exit: 2 });
        }
        return;
    }
    return config.get('appsail');
}
exports.raw = raw;
function getTargetDetails(name) {
    return __awaiter(this, void 0, void 0, function* () {
        const rawTargets = raw();
        const target = rawTargets === null || rawTargets === void 0 ? void 0 : rawTargets.find((tar) => tar.name === name);
        if (!target) {
            return;
        }
        const appConfigPth = (0, project_1.resolveProjectPath)(target.source, constants_1.FILENAME.app_config);
        const isPath = yield fs_1.ASYNC.fileExists(appConfigPth);
        if (!isPath) {
            return Object.assign({ validity: { valid: false, reason: 'Config file not present' } }, target);
        }
        const appConfigJson = yield fs_1.ASYNC.readJSONFile(appConfigPth);
        if (!appConfigJson) {
            return Object.assign({ validity: { valid: false, reason: 'Config file is empty' } }, target);
        }
        return Object.assign(Object.assign(Object.assign({}, target), appConfigJson), { validity: { valid: true } });
    });
}
exports.getTargetDetails = getTargetDetails;
function getAllTargetDetails(throwErr = true) {
    return __awaiter(this, void 0, void 0, function* () {
        const rawTargets = raw();
        if (!rawTargets) {
            if (throwErr) {
                throw new error_1.default('No AppSail targets found');
            }
            return;
        }
        return Promise.all(rawTargets.reduce((resArr, target) => {
            const catalystConfigPth = (0, project_1.resolveProjectPath)(target.source, constants_1.FILENAME.app_config);
            const isPath = fs_1.SYNC.fileExists(catalystConfigPth);
            if (!isPath) {
                resArr.push(Promise.resolve(Object.assign({ validity: {
                        valid: false,
                        reason: 'Config file not present'
                    } }, target)));
                return resArr;
            }
            resArr.push(new Promise((res) => __awaiter(this, void 0, void 0, function* () {
                const configJson = yield fs_1.ASYNC.readJSONFile(catalystConfigPth).catch((err) => (0, logger_1.debug)('Unable to read config file: ', err));
                if (!configJson) {
                    return res(Object.assign({ validity: {
                            valid: false,
                            reason: 'Config file is empty'
                        } }, target));
                }
                return res({
                    appSail: Object.assign(Object.assign({}, target), configJson),
                    validity: { valid: true }
                });
            })));
            return resArr;
        }, []));
    });
}
exports.getAllTargetDetails = getAllTargetDetails;
