// zcatalyst-cli/lib/util_modules/env.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvVariable = exports.envOverride = exports.isPrimaryShell = exports.isWindows = exports.isMac = exports.isCI = void 0;
exports.isCI = process &&
    !!(process.env.CI ||
        process.env.CONTINUOUS_INTEGRATION ||
        process.env.BUILD_NUMBER ||
        process.env.RUN_ID ||
        exports.name ||
        false);
exports.isMac = process && process.platform === 'darwin';
exports.isWindows = process &&
    (process.platform === 'win32' ||
        (process.env.OSTYPE !== undefined && /^(msys|cygwin)$/.test(process.env.OSTYPE)));
function isPrimaryShell() {
    if (!process || process.env.CATALYST_SUB_PROCESS) {
        return false;
    }
    if (!exports.isWindows && process.env.SHLVL) {
        return parseInt(process.env.SHLVL) <= 1;
    }
    return false;
}
exports.isPrimaryShell = isPrimaryShell;
function envOverride(envname, value) {
    var _a;
    if (process.env[envname] !== undefined && ((_a = process.env[envname]) === null || _a === void 0 ? void 0 : _a.length)) {
        return process.env[envname];
    }
    return value;
}
exports.envOverride = envOverride;
function getEnvVariable(variable, fallback) {
    return process.env[variable] || fallback;
}
exports.getEnvVariable = getEnvVariable;
